### PR TITLE
Enhancement/3546 cancel permissions

### PR DESCRIPTION
**Pull request by eugene-manuilov**
## Summary

Addresses issue #3546

## Checklist

- [ ] My code is tested and passes existing unit tests.
- [ ] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 4.7 and PHP 5.6.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [ ] My code has proper inline documentation.
- [ ] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

### CODE REVIEW

**Review by aaemnnosttv — COMMENTED** [review]
Thanks @eugene-manuilov ! This looks good for the most part, although I have some concerns about some of the changes here.

Other than skipping a request, is there a reason why we can't simply submit GA4 settings first (as mentioned in the ACs)?

Would that not accomplish the same thing with a smaller diff while preserving the current messaging?

**Comment by eugene-manuilov** [comment]
> Other than skipping a request, is there a reason why we can't simply submit GA4 settings first (as mentioned in the ACs)?
> Would that not accomplish the same thing with a smaller diff while preserving the current messaging?

@aaemnnosttv if we go this way, then there will be a possibility to get inconsistent settings again:

1. A user selects an existing account, a new UA property, and an existing GA4 property
2. After submitting the form, GA4 settings are saved, but UA settings require an additional scope and it returns an error
3. The setup form shows the permissions popup
4. User cancels it and stops the setup process
5. Now GA4 settings are saved, UA settings aren't

File: assets/js/modules/analytics/components/setup/SetupForm.js [hunk 1]
```diff
@@ -40,47 +40,120 @@ import {
...
 			event.preventDefault();
+
+			const scopes = [];
+
+			if (
+				! hasEditScope &&
+				( uaPropertyID === PROPERTY_CREATE ||
+					uaProfileID === PROFILE_CREATE ||
+					ga4PropertyID === GA4_PROPERTY_CREATE ||
+					ga4WebDataStreamID === WEBDATASTREAM_CREATE )
```
[the remark this answers is not in the document]
Comment (eugene-manuilov) [reply assets/js/modules/analytics/components/setup/SetupForm.js]: Ok, removed UA property checks.

File: assets/js/modules/analytics/components/setup/SetupForm.js [hunk 2]
```diff
@@ -40,47 +40,120 @@ import {
...
+			// If scope not granted, trigger scope error right away. These are
+			// typically handled automatically based on API responses, but
+			// this particular case has some special handling to improve UX.
+			if ( scopes.length > 0 ) {
+				// When state is restored, auto-submit the request again.
+				setValues( FORM_SETUP, { autoSubmit: true } );
+				setPermissionScopeError( {
+					code: ERROR_CODE_MISSING_REQUIRED_SCOPE,
+					message: __(
+						'Additional permissions are required to save Analytics settings.',
```
[the remark this answers is not in the document]
Comment (eugene-manuilov) [reply assets/js/modules/analytics/components/setup/SetupForm.js]: Updated error messages.

**Comment by aaemnnosttv** [comment]
> if we go this way, then there will be a possibility to get inconsistent settings again:
> 
> 1. A user selects an existing account, a new UA property, and an existing GA4 property
> 2. After submitting the form, GA4 settings are saved, but UA settings require an additional scope and it returns an error
> 3. The setup form shows the permissions popup
> 4. User cancels it and stops the setup process
> 5. Now GA4 settings are saved, UA settings aren't

Good point @eugene-manuilov. In that case how about we submit changes for GA4 first if it is creating something, otherwise run it afterwards as it is now. That way, the scope check would always happen before settings are updated but everything else would stay the same?

**Comment by eugene-manuilov** [comment]
> In that case how about we submit changes for GA4 first if it is creating something, otherwise run it afterwards as it is now. That way, the scope check would always happen before settings are updated but everything else would stay the same?

Ok, that makes sense. Updated.

**Review by aaemnnosttv — APPROVED** [review]
LGTM! Thanks @eugene-manuilov

**Pull request merged**